src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Attributes/JsonDerivedTypeAttribute.cs

```diff
@@ -0,0 +1,43 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// When placed on a type declaration, indicates that the specified subtype should be opted into polymorphic serialization.
+    /// </summary>
+    [System.Runtime.Versioning.RequiresPreviewFeatures("Polymorphic serialization is currently in preview.")]
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Interface, AllowMultiple = true, Inherited = false)]
+    public class JsonDerivedTypeAttribute : JsonAttribute
+    {
+        /// <summary>
+        /// Initializes a new attribute with specified parameters.
+        /// </summary>
+        /// <param name="derivedType">A derived type that should be supported in polymorphic serialization of the declared based type.</param>
+        public JsonDerivedTypeAttribute(Type derivedType)
+        {
+            DerivedType = derivedType;
+        }
+
+        /// <summary>
+        /// Initializes a new attribute with specified parameters.
+        /// </summary>
+        /// <param name="derivedType">A derived type that should be supported in polymorphic serialization of the declared base type.</param>
+        /// <param name="typeDiscriminatorId">The type discriminator identifier to be used for the serialization of the subtype.</param>
+        public JsonDerivedTypeAttribute(Type derivedType, string typeDiscriminatorId)
+        {
+            DerivedType = derivedType;
+            TypeDiscriminatorId = typeDiscriminatorId;
+        }
+
+        /// <summary>
+        /// A derived type that should be supported in polymorphic serialization of the declared base type.
+        /// </summary>
+        public Type DerivedType { get; }
```

Comment:
should these 2 be `init` as well?

Reply:
I don't think it needs to, this is a required constructor parameter.